### PR TITLE
Add custom ImagingExtractor for ophys data

### PR DESCRIPTION
**Pull request by weiglszonja**
The current implementation operates on a single .avi file

### CODE REVIEW

File: src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractor.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            self._fps = vc.get_movie_fps()
+
+    def get_frames(self, frame_idxs: ArrayType, channel: int = 0) -> NumpyArray:
+        with self.video_capture_context as vc:
+            rgb_frame = vc.get_movie_frame(frame_number=frame_idxs[0])
+            # Make components of RGB565
+            green_color_data = (rgb_frame[..., 1] >> 2).astype(np.uint16) << 5
+            blue_color_data = (rgb_frame[..., 2] >> 3).astype(np.uint16)
+            # Apply custom conversion to frames : G * 8 + B / 8
+            gray_frame = (green_color_data * 8) + (blue_color_data / 8)
```
Comment (CodyCBakerPhD) [inline src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractor.py:40]: @weiglszonja Remember to cast this back to `uint16`

**Comment by weiglszonja** [comment]
Screenshot 2022-05-27 at 15 18 45

File: src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractor.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    extractor_name = "ScherrerOphysImaging"
+    installed = HAVE_OPENCV
+    installation_mesg = INSTALL_MESSAGE
+
+    def __init__(self, file_path: FilePathType):
+        super().__init__()
+        self.file_path = file_path
+        self.video_capture_context = VideoCaptureContext(str(self.file_path))
+
+        self._num_frames, self._image_size, self._fps = None, None, None
```
Comment (bendichter) [inline src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractor.py:24]: why have this line?

**Comment by weiglszonja** [comment]
How can I add a custom description for the TwoPhotonSeries data?

I found this:
https://github.com/catalystneuro/nwb-conversion-tools/blob/a7cd15f66d1c8239069d964dadf0384ddd7abebe/src/nwb_conversion_tools/tools/roiextractors/roiextractors.py#L285
So the ImagingExtractor should have an attribute with the metadata dict, or is there another way to do this?  


```python
class ScherrerOphysImagingExtractor(ImagingExtractor):
    extractor_name = "ScherrerOphysImaging"
    installed = HAVE_OPENCV
    installation_mesg = INSTALL_MESSAGE
    nwb_metadata = dict(...)
```

**Comment by bendichter** [comment]
@weiglszonja aren't you passing metadata into `run_conversion`? I think it would go in `metadata["Ophys"]["TwoPhotonSeries"]["description"]`

**Comment by CodyCBakerPhD** [comment]
> How can I add a custom description for the TwoPhotonSeries data?

It appears from https://github.com/catalystneuro/nwb-conversion-tools/blob/a7cd15f66d1c8239069d964dadf0384ddd7abebe/src/nwb_conversion_tools/tools/roiextractors/roiextractors.py#L123-L126 that

`metadata["Ophys"]["TwoPhotonSeries"] = [dict(description=your_description))]` If it doesn't yet exist, or

```
Ophys:
    TwoPhotonSeries:
      - description: foobar
```
in the `metadata.yml` file (recommended method).

**Comment by weiglszonja** [comment]
Thanks @CodyCBakerPhD, however
https://github.com/catalystneuro/nwb-conversion-tools/blob/a7cd15f66d1c8239069d964dadf0384ddd7abebe/src/nwb_conversion_tools/tools/roiextractors/roiextractors.py#L98-L99
If I tracked this down correctly what happens is:
`metadata = dict_deep_update(default_ophys_metadata(), metadata) `-- If I have for instance a different description for TwoPhotonSeries, then we update the default one, thats good!
`metadata = safe_update(metadata, get_nwb_imaging_metadata(imaging))` -- But within `get_nwb_imaging_metadata()` we also have `default_ophys_metadata()` so we'll update the description once again, but back to the default one.

Probably omitting this line:
https://github.com/catalystneuro/nwb-conversion-tools/blob/a7cd15f66d1c8239069d964dadf0384ddd7abebe/src/nwb_conversion_tools/tools/roiextractors/roiextractors.py#L184
would fix it, but I'll check it.

**Comment by CodyCBakerPhD** [comment]
@weiglszonja You're right - https://github.com/catalystneuro/nwb-conversion-tools/blob/a7cd15f66d1c8239069d964dadf0384ddd7abebe/src/nwb_conversion_tools/tools/roiextractors/roiextractors.py#L99 should have the order reversed between the two dictionaries, similar to how it's done in other functions throughout that submodule.

File: src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractorinterface.py [hunk 3]
```diff
@@ -20,6 +20,20 @@ def __init__(self, file_path: str):
...
+        metadata = super().get_metadata()
+
+        # Add missing properties for Ophys metadata
+        ophys_metadata = metadata["Ophys"]
+        device_name = ophys_metadata["Device"][0]["name"]
+        imaging_plane = ophys_metadata["ImagingPlane"][0]["name"]
+        if "device" not in ophys_metadata["ImagingPlane"][0].keys():
+            ophys_metadata["ImagingPlane"][0]["device"] = device_name
+        if "imaging_plane" not in ophys_metadata["TwoPhotonSeries"][0].keys():
+            ophys_metadata["TwoPhotonSeries"][0]["imaging_plane"] = imaging_plane
```
Comment (weiglszonja) [inline src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractorinterface.py:33]: Alternatively, changing `default_ophys_metadata()` would also work?
 https://github.com/catalystneuro/nwb-conversion-tools/blob/dacddd4a485bce35cd68c97ef5355aaa84c1ff2c/src/nwb_conversion_tools/tools/roiextractors/roiextractors.py#L39

Comment (CodyCBakerPhD) [reply src/fee_lab_to_nwb/scherrer_ophys/scherrerophysimagingextractorinterface.py:33]: @weiglszonja Quite possibly, yes. That metadata stuff specific to that tool has not been quality-checked in quite some time

**Review by CodyCBakerPhD — APPROVED** [review]

**Pull request merged**